client/core/trade.go

```diff
@@ -334,132 +329,134 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
+
 		match := &matchTracker{
-			id:     mid,
-			prefix: t.Prefix(),
-			trade:  trade,
-			MetaMatch: db.MetaMatch{
-				MetaData: &db.MatchMetaData{
-					Status: order.NewlyMatched,
-					Proof: db.MatchProof{
-						Auth: db.MatchAuth{
-							MatchSig:   msgMatch.Sig,
-							MatchStamp: msgMatch.ServerTime,
-						},
-					},
-					DEX:   t.dc.acct.host,
-					Base:  t.Base(),
-					Quote: t.Quote(),
-				},
-				Match: &order.UserMatch{
-					OrderID:     oid,
-					MatchID:     mid,
-					Quantity:    msgMatch.Quantity,
-					Rate:        msgMatch.Rate,
-					Address:     msgMatch.Address,
-					Status:      order.MatchStatus(msgMatch.Status),
-					Side:        order.MatchSide(msgMatch.Side),
-					FeeRateSwap: feeRateSwap,
-				},
-			},
+			id:        mid,
+			prefix:    t.Prefix(),
+			trade:     trade,
+			MetaMatch: *t.makeMetaMatch(msgMatch),
 		}
 
-		// First check that this isn't a match on its own cancel order. I'm not crazy
-		// about this, but I am detecting this case right now based on the Address
-		// field being an empty string.
-		isCancel := t.cancel != nil && msgMatch.Address == ""
-		if isCancel {
-			match.setStatus(order.MatchComplete)
-			includesCancellation = true
-			log.Infof("maker notification for cancel order received for order %s. match id = %s", oid, msgMatch.MatchID)
-			t.cancel.matches.maker = msgMatch
+		var qty uint64
+		if t.Type() == order.MarketOrderType && !trade.Sell {
+			qty = calc.BaseToQuote(msgMatch.Rate, msgMatch.Quantity)
 		} else {
-			includesTrades = true
-			t.matches[match.id] = match
+			qty = msgMatch.Quantity
+		}
+		newFill += qty
+
+		log.Infof("Starting negotiation for match %v for order %v with swap fee rate = %v, quantity = %v",
+			mid, oid, match.MetaMatch.Match.FeeRateSwap, qty)
+		match.SetStatus(order.NewlyMatched) // these must be new matches
+		newTrackers = append(newTrackers, match)
+	}
+
+	// Now that each Match in msgMatches has been validated, store them in the
+	// trackedTrade and the DB, and update each order's status and fill amount.
+	// First record any cancel order Match and update order status.
+	if cancelMatch != nil {
+		t.cancel.matches.maker = cancelMatch // taker is stored via processCancelMatch before negotiate
+		// Set the order status for both orders.
+		t.metaData.Status = order.OrderStatusCanceled
+		t.db.UpdateOrderStatus(t.cancel.ID(), order.OrderStatusExecuted)
+		// Store a completed maker cancel match in the DB.
+		makerCancelMeta := t.makeMetaMatch(cancelMatch)
+		makerCancelMeta.SetStatus(order.MatchComplete)
+		err := t.db.UpdateMatch(makerCancelMeta)
+		if err != nil {
+			return fmt.Errorf("failed to update match in db: %w", err)
 		}
+		// Store a completed taker cancel match too.
+		takerCancelMeta := t.makeMetaMatch(t.cancel.matches.taker)
+		takerCancelMeta.SetStatus(order.MatchComplete)
+		err = t.db.UpdateMatch(takerCancelMeta)
+		if err != nil {
+			return fmt.Errorf("failed to update match in db: %w", err)
+		}
```

Comment:
Why not handle this in `processCancelMatch`?

Reply:
I wanted to update the db at roughly the same time, as the comment at the end `processCancelMatch` says without any real justification though.  `processCancelMatch` and `negotiate` are run in sequence, so I'll just do it in `processCancelMatch`.